Implement the reference Fluent Syntax parser

stasm opened the pull request:
The _kombi_ parser started out as an experiment with LL parser combinators two weeks ago when I met with @Pike in Berlin. The initial prototype lived in https://github.com/stasm/fluent-kombi. I refined it to a point where it's ready to become the reference parser implementation.

This PR removes `diagram.xhtml` because the EBNF generated from `syntax/grammar.mjs` contains RegExps, which are not parsed by the [Railroad Diagrams Generator](http://www.bottlecaps.de/rr/ui). I expect we'll use more RegExps in the future for performance reasons, but for now I deliberately only used it in a few places (mostly as `/.*/`). The PR also replaces the ASDL with `syntax/ast.mjs` defining the AST data structures and `syntax/abstract.mjs` defining how to transforms the CST (concrete parse tree, the direct output of the parser) into the AST.

Unfortunately, I wasn't yet able to use the [fixtures from `fluent-syntax`](https://github.com/projectfluent/fluent.js/tree/master/fluent-syntax/test) because the reference parser doesn't produce any annotations and doesn't add spans to AST nodes. I'd like to add both of these features in the future. I added a new set of fixtures and in some cases I copied the FTL fixtures from `fluent-syntax` and generated new JSON AST for them. The current coverage of the tests is similar to the `fixtures_structure` one's from `fluent-syntax`.

Pike commented on syntax/grammar.mjs:
```diff
@@ -0,0 +1,495 @@
+    .map(to_object)
+    .map(into(FTL.Term)));
+
+let Attribute = defer(() =>
```
How about moving that until after the pattern block? I find it's position here to be a bit of a break in reading the grammar and the ebnf.

stasm replied:
Good point. I'll move it into it's own comment group right above `Pattern`, because `Attribute` references it as well and I'd like to keep all `Expressions` right below `Pattern`.

Pike commented on syntax/grammar.mjs:
```diff
@@ -0,0 +1,495 @@
+    .map(into(FTL.ResourceComment)));
+
+let Junk = defer(() =>
+    repeat1(
```
Looking at how this renders and reads, what do you think about making Junk just single line in the syntax, and glue Junk nodes together in in abstract.mjs?

That would Junk just make "eat the rest of the line", right?

stasm replied:
Interesting idea, that would work, yes. It would only avoid the `repeat1` though, right? I think I'd prefer to keep it as it is right now. In fact, I already have a few ideas for changes that will reduce the amount of post-processing in `abstract.mjs`. If something can be expressed in the grammar, my current thinking is to do it there.

Pike replied:
I expect it to actually remove all the needs for `not(Entry)`

stasm replied:
That would be a win. I'll try it, thanks!

stasm replied:
I tried it out and it works well except that `Junk` doesn't now include trailing blank lines (if any). Which might actually make it a bit more like other `Entries`.

Pike replied:
so if we'd join them afterwards, we'd have to pad our own newline? Or add top-level whitespace to the `Resource` production?

stasm replied:
Post-processing Junk changes how the following is parsed:

```
junk_line1\n
\n
junk_line2\n
\n
```

Before it was a single `Junk` entry which included all 4 `\n`. When we join junk lines in post-processing, those blank_lines don't make it there because they are dropped from the CST before `map(into(FTL.Resource))` is called. In a naive joining approach, the new single `Junk` would now only contain `junk_line1\njunk_line2`. Note the missing middle and trailing `\n`

One option is to let blank lines make it to the `FTL.Resource` construction and strip them there as the last post-processing step. This prevents the two junk lines above from being joined because there aren't adjacent now. The result is that we end up with _two_ `Junk` entries now, each for a contiguous block of junk lines.

I implemented this behavior in a commit which I'll push in a few.

Pike commented on syntax/grammar.mjs:
```diff
@@ -0,0 +1,495 @@
+        Function.as("callee"),
+        char("("),
+        between(
+            maybe(inline_space),
```
I think this reads better without the `between inline_space`. We have that around each argument anyway.

stasm replied:
We need at least the `maybe(inline_space)` part to handle the case of no arguments: `FUNCTION( <whitespace here>     )`.

Pike commented on syntax/grammar.mjs:
```diff
@@ -0,0 +1,496 @@
+    .map(into(FTL.Variant))
+    .map(mutate({default: true})));
+
+let SelectorExpression = defer(() =>
```
I think this would read better if it was right after `SelectExpression`.

Pike commented on syntax/grammar.mjs:
```diff
@@ -0,0 +1,496 @@
+        repeat(Variant))
+    .map(flatten(1)));
+
+let VariantName = defer(() =>
```
`VariantName` and `VariantKey` threw me off here coming before `Variant` and `DefaultVariant`. Want to move them down?

Pike commented on syntax/grammar.mjs:
```diff
@@ -0,0 +1,496 @@
+    .map(to_object)
+    .map(into(FTL.AttributeExpression)));
+
+let TermAttributeExpression = defer(() =>
```
Move this below `SelectorExpression`, and in particular out of inline expressions? As it's not one ;-)

Pike commented:
I think there are a couple of things that'd be good to put into the changelog, and probably also the final commit message. I'd call out

* explicit addition of `Junk`
* `Comment` being explicitly bound to `Message` and `Term` if adjacent
* obviously the shift of roles between adsl and mjs

I've made a pass through the ebnf in full now, the changes look good.

I'm still concerned about the plan to drop the ebnf. Not sure if that's my brain only, but I benefit significantly from the consice notion. I'll go through the exercise of reading grammar.mjs again tomorrow. Adding that comment here, because the role of the ebnf should also be part of the changelog entry, I guess.

Pike commented:
Now, onwards to the role of `grammar.mjs` as a reference parser.

I wonder if we should refactor `regular_char` and `text_char` alongside with `word` and `text_cont` to avoid the exceptions into an `either`. My guess is that many parser architectures will struggle to implement that as is, and we might be better off giving an example of how to implement it right?

Maybe with a full `ordenary_char` as `regular_char without any jazz`, and then just add our special chars explicitly where allowed?

Also, it might be easier to read if we added/removed character groups rather than single chars.

Side note, I wonder if we should have both `char()` and `string()`. I consider the difference to be an implementation detail, and the implementation of `string` is sadly one that I blame for many of the perf problems of the current python parser. The profile I had was showing deep stacks of single char checks. This being reference impl and all, we could just suggest to implement string checks as full string checks.

stasm commented:
> I'm still concerned about the plan to drop the ebnf. Not sure if that's my brain only, but I benefit significantly from the consice notion. I'll go through the exercise of reading grammar.mjs again tomorrow. Adding that comment here, because the role of the ebnf should also be part of the changelog entry, I guess.

I don't think we should drop it entirely, sorry if I didn't make it clear before. I think we should base our discussions on `grammar.mjs` because it allows ideas to be tested but continue to use the EBNF as quickly-look-something-up reference and as a easily digestible grammar cheatsheet.

(I didn't write the whole EBNF serialization just for the sake of this PR 😄. I want to keep using it going forward).

stasm commented:
Can we move all of your feedback from https://github.com/projectfluent/fluent/pull/103#issuecomment-389487454 re. `text_char`, `regular_char`, character classes and `string` vs. `char` into follow-ups? They are valuable improvements but I wouldn't like to block landing this.

stasm commented on lib/combinators.mjs:
```diff
@@ -0,0 +1,139 @@
+
+export function defer(fn) {
+    // Parsers might be defined as defer(() => parser) to avoid cyclic
+    // dependecies.
```
We still have a few parsers down at the bottom of `grammar.mjs` which don't use `defer`.

stasm commented on test/fixtures/select_expressions.ftl:
```diff
@@ -0,0 +1,25 @@
+variant-list =
+    {
+       *[key] value
+    }
```
I'd prefer to stick to the current EBNF as much as possible, even if it means implementing something we already agreed should go away.

It's a bad practice to do this in a Message, yes. I'll add semantics checks later, after this lands. That's #99.

stasm commented on test/fixtures/escaped_characters.ftl:
```diff
@@ -0,0 +1,5 @@
+backslash = Value with \\ (an escaped backslash)
+closing-brace = Value with \{ (a closing brace)
+unicode-escape = \u0041
+escaped-unicode = \\u0041
```
Ouch, thanks for catching that. I'll fix it by storing the escaped `\` in the AST.

stasm commented:
Thanks again, @Pike!

> I think that in 0.5, we allow for multline argument lists, with the infamous `__?` being an optional break-indent in https://github.com/projectfluent/fluent/blob/69ecabcc06cf6f5b3654c439a9defc8575744313/spec/fluent.ebnf#L77-L78.

Right, good catch. It wouldn't be beneficial to try to match the old EBNF because it's wrong. The `__?` doesn't allow for any inline space before the `break_indent`. I added a new `space_indent` production defined as `(inline_space break_indent?) | (inline_space? break_indent)` to fix this. I changed `CallExpression` and `Argument` to use it.

We should fix the same `__?` in `Placeable` but that's a bit harder because we don't want to consume the `break_indent` too quickly in case it's a `variant_list` (which must start with a `break_indent` too). I left only `inline_space` inside of `Placeable` for now. It's still an improvement over the old EBNF (which was broken) and we'll fix it in #102 by separating variant lists out of Placeables completely.

> The next issue is that `\u1234` and `\\u1234` result in the same AST. That's because the unescaping of `\` is assymetric between the unicode escape and `\\`, `\{`, etc. I suggest to not unescape at the syntax or ast level. Maybe we should have an AST node `Backtick`?

I agree with your earlier comment: let's store the special characters in their escaped form in the AST.

> I've also looked at the use of `element_at`, even before I hit the unescaping issue. In the non-escaping cases, would using `.abstract` and `.map(keep_abstract)` be more consistent than `element_at(2)`? This is used to skip whitespace etc in the AST a twice.

Because of the above, a simple `map(join)` does the trick :)

> Lastly, I hit a technical snag trying to check out different branches by a locally modified package-lock.json, in our case, I don't think we should commit it? Or actually pin versions so that local installs don't conflict?

`package-lock.json` isn't committed on master, so I'm not sure what kind of conflict you ran into. In any case, I think we _should_ commit it and yes, we should also pin versions of dependencies exactly (which we do) (unless I misunderstood?).

> And I forgot that I asked about the changelog entries, but maybe we do them in a follow-up.

I'll add it in this PR and then probably edit them again when publishing the 0.6 version. Thanks for the reminder!

Pike commented on syntax/grammar.mjs:
```diff
@@ -0,0 +1,502 @@
+        either(
+            // Order matters!
+            BlockExpression,
+            InlineExpression),
```
following up on an earlier comment, should this be `either(BE, IE).abstract` and s/element_at/keep_abstract/ ?

Pike replied:
sounds good. I had the feeling you'd like to keep the maybe() top-level ;-), the way it renders in the ebnf is a good supportive argument

stasm replied:
Is this comment about the convention or is there some other reason to prefer `keep_abstract`? The reason why I used `element_at` here is that it returns a scalar value so I can then use `into` rather than `list_into` which made a little bit more sense to me semantically. There's only one thing that we care about inside of `Placeable`. I'm not overly attached to this reasoning, however, so I'm happy to change it if you feel it will improve the readability.

stasm commented on syntax/grammar.mjs:
```diff
@@ -0,0 +1,502 @@
+            maybe(break_indent)),
+        sequence(
+            maybe(inline_space),
+            break_indent));
```
I'd like to keep using the `maybe()` combinator because it serializes as `?` in the EBNF. Your version of `space_indent` is how I originally implemented it :) I changed it to remove the redyndancy of nested `maybes`. If you're OK with it, I'll go with `maybe(space_indent)` and your implementation of `space_indent`.
